modules/ml/src/main/java/org/apache/ignite/ml/tree/randomforest/RandomForestTrainer.java

```diff
@@ -17,45 +17,815 @@
+import org.apache.ignite.ml.dataset.primitive.builder.context.EmptyContextBuilder;
+import org.apache.ignite.ml.dataset.primitive.context.EmptyContext;
+import org.apache.ignite.ml.math.functions.IgniteBiFunction;
+import org.apache.ignite.ml.math.functions.IgniteFunction;
+import org.apache.ignite.ml.math.primitives.vector.Vector;
+import org.apache.ignite.ml.trainers.DatasetTrainer;
+import org.apache.ignite.ml.tree.randomforest.data.FeaturesCountSelectionStrategies;
+import org.apache.ignite.ml.tree.randomforest.data.NodeId;
+import org.apache.ignite.ml.tree.randomforest.data.NodeSplit;
+import org.apache.ignite.ml.tree.randomforest.data.TreeNode;
+import org.apache.ignite.ml.tree.randomforest.data.impurity.GiniHistogram;
+import org.apache.ignite.ml.tree.randomforest.data.impurity.ImpurityComputer;
 
 /**
- * Abstract random forest trainer.
+ * Class represents a realization of Random Forest algorithm. Main idea of this realization is that at each
+ * learning iteration it try to aggregate statistics on impurity for each corner nodes in each trees in
+ * random forest. It require one map-reduce operation over learning dataset. After such aggregation
+ * the algorithm select split points for each corner nodes of create leaf nodes. The algorithm stops when
+ * there is no splitting for nodes in trees. At last stage the algorithm aggregate statistics on labels for
+ * leaf nodes in one map-reduce step and sets values to leafs based these statistics.
+ *
+ * @param <LeafStatsAggregator> Type of statistics aggregator for leaf values computing.
+ * @param <S> Type of impurity computer specific for algorithm.
+ * @param <T> Type of child of RandomForestTrainer using in with-methods.
  */
-public abstract class RandomForestTrainer extends BaggingModelTrainer {
-    /** Max decision tree deep. */
-    protected final int maxDeep;
+public abstract class RandomForestTrainer<LeafStatsAggregator, S extends ImpurityComputer<BootstrappedVector, S>, T extends RandomForestTrainer<LeafStatsAggregator, S, T>>
+    extends DatasetTrainer<ModelsComposition, Double> {
+    /** Bucket size factor. */
+    public static final double BUCKET_SIZE_FACTOR = (1 / 10.0);
+
+    /** Count of trees. */
+    private int countOfTrees;
+
+    /** Subsample size. */
+    private double subsampleSize;
+
+    /** Max depth. */
+    private int maxDepth;
+
+    /** Min impurity delta. */
+    private double minImpurityDelta;
+
+    /** Features Meta. */
+    private List<FeatureMeta> meta;
+    private int featuresPerTree;
+
+    /** Random generator. */
+    private Random random;
+
+    /** Seed. */
+    private long seed;
+
+    /** Nodes to learn selection strategy. */
+    private IgniteFunction<Queue<TreeNode>, List<TreeNode>> nodesToLearnSelectionStrgy;
+
+    /**
+     * Create an instance of RandomForestTrainer.
+     *
+     * @param meta Features Meta.
+     */
+    public RandomForestTrainer(List<FeatureMeta> meta) {
+        this.meta = meta;
```

Comment:
it seems like a section with default parameters. The bad idea to do that in constructor. Please, move this functionality to the field declaration section.